File: apps/ios/GalaxySSI/AgentEvalBenchmarkModels.swift
import Foundation

enum AgentBenchmarkDimension: String, Codable, CaseIterable, Identifiable {
  case taskQuality = "task_quality"
  case planningAndTools = "planning_and_tools"
  case iosWorld = "ios_world"
  case immediateMemory = "immediate_memory"
  case longTermMemory = "long_term_memory"
  case recovery
  case multiAgent = "multi_agent"

  var id: String { rawValue }
}

struct AgentBenchmarkExpectation: Codable, Equatable {
  var requiredOutputPatterns: [String] = []
  var requiredJsonFields: [String: String] = [:]
  var forbiddenOutputPatterns: [String] = []
  var minimumOutputCharacters: Int = 1
  var minimumPlanEvents: Int = 0
  var minimumToolReceipts: Int = 0
  var minimumVerifiedSources: Int = 0
  var minimumDistinctAgents: Int = 1
  var minimumHandoffs: Int = 0
  var requiredEvidence: Set<AgentOutcomeEvidenceKind> = [.finalResponse]
  var requiredCondition: AgentEvalCondition = .normal
  var memoryHorizonDays: Int = 0
  var iosWorldTaskId: String = ""
  var requireIOSObservedValuesInOutput = false
}

enum AgentBenchmarkReadinessStatus: String, Codable, CaseIterable {
  case ready
  case waiting
  case blocked
}

struct AgentBenchmarkCaseReadiness: Codable, Equatable {
  var caseId: String
  var status: AgentBenchmarkReadinessStatus
  var reasonCode: String = ""
  var eligibleAtMillis: Int64 = 0
}

struct AgentBenchmarkCase: Codable, Equatable, Identifiable {
  var id: String
  var dimension: AgentBenchmarkDimension
  var title: String
  var prompt: String
  var expectation: AgentBenchmarkExpectation
  var critical = true

  var taggedPrompt: String {
    var value = "[evalops:\(id)] \(prompt.trimmingCharacters(in: .whitespacesAndNewlines))"
    if !expectation.iosWorldTaskId.isEmpty {
      value += " [iosworld:\(expectation.iosWorldTaskId)]"
    }
    return value
  }
}

struct AgentBenchmarkSuite: Codable, Equatable, Identifiable {
  var id: String
  var version: String
  var title: String
  var cases: [AgentBenchmarkCase]
  var targetPassRate = 0.90
  var minimumTaskCount = 50
  var maximumTaskCount = 100
  var minimumRepetitions = 3
  var maximumRepetitions = 10

  init(
    id: String,
    version: String,
    title: String,
    cases: [AgentBenchmarkCase],
    targetPassRate: Double = 0.90,
    minimumTaskCount: Int = 50,
    maximumTaskCount: Int = 100,
    minimumRepetitions: Int = 3,
    maximumRepetitions: Int = 10
  ) {
    precondition((minimumTaskCount...maximumTaskCount).contains(cases.count))
    precondition(Set(cases.map(\.id)).count == cases.count)
    self.id = id
    self.version = version
    self.title = title
    self.cases = cases
    self.targetPassRate = min(max(targetPassRate, 0), 1)
    self.minimumTaskCount = minimumTaskCount
    self.maximumTaskCount = maximumTaskCount
    self.minimumRepetitions = minimumRepetitions
    self.maximumRepetitions = maximumRepetitions
  }

  func benchmarkCase(id: String) -> AgentBenchmarkCase? {
    cases.first { $0.id == id }
  }
}

struct AgentBenchmarkResourceSnapshot: Codable, Equatable, Identifiable {
  var resourceId: String
  var displayName: String
  var providerId: String
  var modelId: String
  var adapterType: String
  var capabilitiesHash: String
  var id: String { resourceId }
}

enum AgentBenchmarkSessionStatus: String, Codable, CaseIterable {
  case running
  case completed
  case cancelled
}

struct AgentBenchmarkSession: Codable, Equatable, Identifiable {
  var id: String = UUID().uuidString
  var suiteId: String
  var suiteVersion: String
  var appVersionName: String
  var appBuildNumber: String
  var deviceModel: String
  var systemVersion: String
  var repetitions: Int
  var targetPassRate: Double
  var caseIds: [String]
  var resources: [AgentBenchmarkResourceSnapshot]
  var resourceIdsByCase: [String: [String]]
  var campaignIdsByCase: [String: String]
  var teamResourceIdsByCase: [String: [String]] = [:]
  var readinessByCase: [String: AgentBenchmarkCaseReadiness] = [:]
  var allocationProfile = "codex_90_deepseek_10"
  var status: AgentBenchmarkSessionStatus = .running
  var createdAtMillis: Int64 = AgentEvalClock.nowMillis()
  var updatedAtMillis: Int64 = AgentEvalClock.nowMillis()

  var scheduledCaseIds: [String] {
    readinessByCase.isEmpty ? caseIds : caseIds.filter { readinessByCase[$0]?.status == .ready }
  }

  var expectedTrials: Int {
    scheduledCaseIds.reduce(0) { $0 + (resourceIdsByCase[$1]?.count ?? 0) } * repetitions
  }
}

extension AgentBenchmarkSession {
  private enum CodingKeys: String, CodingKey {
    case id, suiteId, suiteVersion, appVersionName, appBuildNumber, deviceModel, systemVersion
    case repetitions, targetPassRate, caseIds, resources, resourceIdsByCase, campaignIdsByCase
    case teamResourceIdsByCase, readinessByCase, allocationProfile, status, createdAtMillis, updatedAtMillis
  }

  init(from decoder: Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)
    id = try values.decodeIfPresent(String.self, forKey: .id) ?? UUID().uuidString
    suiteId = try values.decodeIfPresent(String.self, forKey: .suiteId) ?? ""
    suiteVersion = try values.decodeIfPresent(String.self, forKey: .suiteVersion) ?? ""
    appVersionName = try values.decodeIfPresent(String.self, forKey: .appVersionName) ?? ""
    appBuildNumber = try values.decodeIfPresent(String.self, forKey: .appBuildNumber) ?? ""
    deviceModel = try values.decodeIfPresent(String.self, forKey: .deviceModel) ?? ""
    systemVersion = try values.decodeIfPresent(String.self, forKey: .systemVersion) ?? ""
    repetitions = try values.decodeIfPresent(Int.self, forKey: .repetitions) ?? 1
    targetPassRate = try values.decodeIfPresent(Double.self, forKey: .targetPassRate) ?? 0.90
    caseIds = try values.decodeIfPresent([String].self, forKey: .caseIds) ?? []
    resources = try values.decodeIfPresent([AgentBenchmarkResourceSnapshot].self, forKey: .resources) ?? []
    resourceIdsByCase = try values.decodeIfPresent([String: [String]].self, forKey: .resourceIdsByCase) ?? [:]
    campaignIdsByCase = try values.decodeIfPresent([String: String].self, forKey: .campaignIdsByCase) ?? [:]
    teamResourceIdsByCase = try values.decodeIfPresent([String: [String]].self, forKey: .teamResourceIdsByCase) ?? [:]
    readinessByCase = try values.decodeIfPresent([String: AgentBenchmarkCaseReadiness].self, forKey: .readinessByCase) ?? [:]
    allocationProfile = try values.decodeIfPresent(String.self, forKey: .allocationProfile) ?? "codex_90_deepseek_10"
    status = try values.decodeIfPresent(AgentBenchmarkSessionStatus.self, forKey: .status) ?? .running
    createdAtMillis = try values.decodeIfPresent(Int64.self, forKey: .createdAtMillis) ?? AgentEvalClock.nowMillis()
    updatedAtMillis = try values.decodeIfPresent(Int64.self, forKey: .updatedAtMillis) ?? createdAtMillis
  }
}

struct AgentBenchmarkTrialResult: Codable, Equatable, Identifiable {
  var id: String = UUID().uuidString
  var sessionId: String
  var caseId: String
  var campaignId: String
  var trialId: String
  var runId: String
  var resourceId: String
  var repetition: Int
  var passed: Bool
  var verified: Bool
  var failureReasons: [String]
  var durationMillis: Int64
  var reportedCostMicros: Int64
  var batteryDeltaPercent: Int
  var peakThermalStatus: Int
  var completedAtMillis: Int64 = AgentEvalClock.nowMillis()
}

struct AgentBenchmarkMetric: Codable, Equatable, Identifiable {
  var dimension: AgentBenchmarkDimension?
  var taskCount: Int
  var coveredTaskCount: Int
  var expectedTrials: Int
  var completedTrials: Int
  var verifiedTrials: Int
  var passAt1: Double?
  var passPowerK: Double?
  var averageLatencyMillis: Int64
  var averageReportedCostMicros: Int64
  var averageBatteryDeltaPercent: Double
  var peakThermalStatus: Int
  var qualified: Bool
  var targetMet: Bool
  var passedTrials: Int = 0
  var capabilityFailureTrials: Int = 0
  var infrastructureFailureTrials: Int = 0
  var waitingForRealConditionTrials: Int = 0
  var evaluableTrials: Int = 0
  var evaluableTaskCount: Int = 0
  var certificationComplete = false
  var plannedTrials: Int = 0
  var notExecutedTrials: Int = 0
  var blockedTrials: Int = 0
  var certificationCoverage: Double? = nil
  var id: String { dimension?.rawValue ?? "overall" }
}

struct AgentBenchmarkTrialEvidence: Codable, Equatable, Identifiable {
  var caseId: String
  var caseTitle: String
  var dimension: AgentBenchmarkDimension
  var resourceName: String
  var repetition: Int
  var classification: AgentBenchmarkTrialClassification
  var failureReasons: [String]
  var rawOutput: String
  var planEventCount: Int
  var toolReceipts: [String]
  var iosWorldEvidence: [String]
  var runId: String

  var id: String { runId }
}

struct AgentBenchmarkResourceScore: Codable, Equatable, Identifiable {
  var resource: AgentBenchmarkResourceSnapshot
  var overall: AgentBenchmarkMetric
  var dimensions: [AgentBenchmarkMetric]
  var id: String { resource.resourceId }
}

struct AgentBenchmarkScorecard: Codable, Equatable {
  var session: AgentBenchmarkSession
  var overall: AgentBenchmarkMetric
  var dimensions: [AgentBenchmarkMetric]
  var resources: [AgentBenchmarkResourceScore]
  var generatedAtMillis: Int64 = AgentEvalClock.nowMillis()
}

struct AgentBenchmarkProgress: Codable, Equatable {
  var completedTrials: Int
  var expectedTrials: Int
  var completedCampaigns: Int
  var totalCampaigns: Int
  var terminal: Bool
}

struct AgentBenchmarkAllocation: Equatable {
  var resources: [AgentRegistration]
  var resourceIdsByCase: [String: [String]]
  var teamResourceIdsByCase: [String: [String]] = [:]
}

struct AgentBenchmarkError: LocalizedError, Equatable {
  var message: String
  var errorDescription: String? { message }
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/AgentEvalBenchmarkModels.kt
package com.galaxyssi.chat

import java.util.UUID

enum class AgentBenchmarkDimension(val wireValue: String) {
    TASK_QUALITY("task_quality"),
    PLANNING_AND_TOOLS("planning_and_tools"),
    ANDROID_WORLD("android_world"),
    IMMEDIATE_MEMORY("immediate_memory"),
    LONG_TERM_MEMORY("long_term_memory"),
    RECOVERY("recovery"),
    MULTI_AGENT("multi_agent")
}

data class AgentBenchmarkExpectation(
    val requiredOutputPatterns: List<String> = emptyList(),
    val requiredJsonFields: Map<String, String> = emptyMap(),
    val forbiddenOutputPatterns: List<String> = emptyList(),
    val minimumOutputChars: Int = 1,
    val minimumPlanEvents: Int = 0,
    val minimumToolReceipts: Int = 0,
    val minimumVerifiedSources: Int = 0,
    val minimumDistinctAgents: Int = 1,
    val minimumHandoffs: Int = 0,
    val requiredEvidence: Set<AgentOutcomeEvidenceKind> = setOf(AgentOutcomeEvidenceKind.FINAL_RESPONSE),
    val requiredCondition: AgentEvalCondition = AgentEvalCondition.NORMAL,
    val memoryHorizonDays: Int = 0,
    val androidWorldTaskId: String = "",
    val requireAndroidObservedValuesInOutput: Boolean = false
)

data class AgentBenchmarkCase(
    val id: String,
    val dimension: AgentBenchmarkDimension,
    val title: String,
    val prompt: String,
    val expectation: AgentBenchmarkExpectation,
    val critical: Boolean = true
) {
    val taggedPrompt: String get() = buildString {
        append("[evalops:").append(id).append("] ").append(prompt.trim())
        if (expectation.androidWorldTaskId.isNotBlank()) {
            append(" [androidworld:").append(expectation.androidWorldTaskId).append(']')
        }
    }
}

data class AgentBenchmarkSuite(
    val id: String,
    val version: String,
    val title: String,
    val cases: List<AgentBenchmarkCase>,
    val targetPassRate: Double = 0.90,
    val minimumTaskCount: Int = 50,
    val maximumTaskCount: Int = 100,
    val minimumRepetitions: Int = 3,
    val maximumRepetitions: Int = 10
) {
    init {
        require(cases.size in minimumTaskCount..maximumTaskCount)
        require(cases.map(AgentBenchmarkCase::id).distinct().size == cases.size)
        require(targetPassRate in 0.0..1.0)
    }

    fun case(id: String): AgentBenchmarkCase? = cases.firstOrNull { it.id == id }
}

data class AgentBenchmarkResourceSnapshot(
    val resourceId: String,
    val displayName: String,
    val providerId: String,
    val modelId: String,
    val adapterType: String,
    val capabilitiesHash: String
)

enum class AgentBenchmarkReadinessStatus { READY, WAITING, BLOCKED }

data class AgentBenchmarkCaseReadiness(
    val caseId: String,
    val status: AgentBenchmarkReadinessStatus,
    val reasonCode: String = "",
    val eligibleAtMillis: Long = 0L
)

enum class AgentBenchmarkSessionStatus { RUNNING, COMPLETED, CANCELLED }

data class AgentBenchmarkSession(
    val id: String = UUID.randomUUID().toString(),
    val suiteId: String,
    val suiteVersion: String,
    val appVersionName: String,
    val appVersionCode: Long,
    val deviceModel: String,
    val repetitions: Int,
    val targetPassRate: Double,
    val caseIds: List<String>,
    val resources: List<AgentBenchmarkResourceSnapshot>,
    val resourceIdsByCase: Map<String, List<String>>,
    val campaignIdsByCase: Map<String, String>,
    val teamResourceIdsByCase: Map<String, List<String>> = emptyMap(),
    val readinessByCase: Map<String, AgentBenchmarkCaseReadiness> = emptyMap(),
    val allocationProfile: String = "codex_90_deepseek_10",
    val status: AgentBenchmarkSessionStatus = AgentBenchmarkSessionStatus.RUNNING,
    val createdAtMillis: Long = System.currentTimeMillis(),
    val updatedAtMillis: Long = createdAtMillis
) {
    val scheduledCaseIds: List<String> get() = if (readinessByCase.isEmpty()) {
        caseIds
    } else {
        caseIds.filter { readinessByCase[it]?.status == AgentBenchmarkReadinessStatus.READY }
    }
    val expectedTrials: Int get() = scheduledCaseIds.sumOf {
        resourceIdsByCase[it].orEmpty().size
    } * repetitions
}

data class AgentBenchmarkTrialResult(
    val id: String = UUID.randomUUID().toString(),
    val sessionId: String,
    val caseId: String,
    val campaignId: String,
    val trialId: String,
    val runId: String,
    val resourceId: String,
    val repetition: Int,
    val passed: Boolean,
    val verified: Boolean,
    val failureReasons: List<String>,
    val durationMillis: Long,
    val reportedCostMicros: Long,
    val batteryDeltaPercent: Int,
    val peakThermalStatus: Int,
    val completedAtMillis: Long = System.currentTimeMillis()
)

data class AgentBenchmarkMetric(
    val dimension: AgentBenchmarkDimension?,
    val taskCount: Int,
    val coveredTaskCount: Int,
    val expectedTrials: Int,
    val completedTrials: Int,
    val verifiedTrials: Int,
    val passAt1: Double?,
    val passPowerK: Double?,
    val averageLatencyMillis: Long,
    val averageReportedCostMicros: Long,
    val averageBatteryDeltaPercent: Double,
    val peakThermalStatus: Int,
    val qualified: Boolean,
    val targetMet: Boolean,
    val passedTrials: Int = 0,
    val capabilityFailureTrials: Int = 0,
    val infrastructureFailureTrials: Int = 0,
    val waitingForRealConditionTrials: Int = 0,
    val evaluableTrials: Int = 0,
    val evaluableTaskCount: Int = 0,
    val certificationComplete: Boolean = false,
    val plannedTrials: Int = expectedTrials,
    val notExecutedTrials: Int = 0,
    val blockedTrials: Int = 0,
    val certificationCoverage: Double? = null
)

data class AgentBenchmarkTrialEvidence(
    val caseId: String,
    val caseTitle: String,
    val dimension: AgentBenchmarkDimension,
    val resourceName: String,
    val repetition: Int,
    val classification: AgentBenchmarkTrialClassification,
    val failureReasons: List<String>,
    val rawOutput: String,
    val planEventCount: Int,
    val toolReceipts: List<String>,
    val androidWorldEvidence: List<String>,
    val runId: String
)

data class AgentBenchmarkResourceScore(
    val resource: AgentBenchmarkResourceSnapshot,
    val overall: AgentBenchmarkMetric,
    val dimensions: List<AgentBenchmarkMetric>
)

data class AgentBenchmarkScorecard(
    val session: AgentBenchmarkSession,
    val overall: AgentBenchmarkMetric,
    val dimensions: List<AgentBenchmarkMetric>,
    val resources: List<AgentBenchmarkResourceScore>,
    val generatedAtMillis: Long = System.currentTimeMillis()
)

data class AgentBenchmarkProgress(
    val completedTrials: Int,
    val expectedTrials: Int,
    val completedCampaigns: Int,
    val totalCampaigns: Int,
    val terminal: Boolean
)
